mlflow/gateway/fluent.py

```diff
@@ -46,7 +46,7 @@ def pagination_wrapper_func(_, next_page_token):
 
 
 @experimental
-def create_route(name: str, route_type: str, model: Dict[str, Any]) -> Route:
+def create_route(name: str, model: Dict[str, Any], route_type: Optional[str] = None) -> Route:
```

Comment:
This looks like a breaking change. Is it ok?

Reply:
Yeah I flagged it as a breaking change for that reason. The alternative (if we want to preserve parameter order) would be to make the model parameter optional as well and check for it in the function

Reply:
Another alternative is:

```
def create_route(*, name: str, route_type: Optional[str] = None, model: Dict[str, Any]) -> Route:
```

Reply:
or we can use a trick like this:

```python
import functools
import warnings


def decorator(f):
    @functools.wraps(f)
    def decorated(*args, **kwargs):
        if "route_type" in kwargs:
            return f(*args, **kwargs)
        else:
            [name, route_type, *rest] = args
            warnings.warn("Swap model and route_type", DeprecationWarning)
            return f(name, *rest, **kwargs, route_type=route_type)

    return decorated


def create_route(name, model, route_type=None):
    assert isinstance(model, dict), "model must be a dict"
    print(name, model, route_type)


name = "my-route"
route_type = "llm/v1/completions"
model = {
    "name": "question-answering-bot",
    "provider": "openai",
    "openai_config": {
        "openai_api_key": "foo",
    },
}

try:
    create_route(name, route_type, model)
except AssertionError as e:
    print("Failed as expected", e)
else:
    raise AssertionError("AssertionError not raised")

non_breaking_create_route = decorator(create_route)
non_breaking_create_route(name, route_type, model)  # all positional
non_breaking_create_route(name, route_type=route_type, model=model)  # route_type as kwarg

```

Reply:
@dbczumar @BenWilson2 thoughts on which approach we should take?

Reply:
Could try using a sentinel value to keep the signature the same?

```python
_sentinel = object()

@experimental
def create_route(self, name: str, route_type=_sentinel, model: Dict[str, Any]) -> Route:
    """
    docstring contents
    """
    if not self._is_databricks_host():
        raise MlflowException(
            "The create_route API is only available when running within "
            "Databricks. Route creation is handled through creating a "
            "configuration YAML file during startup or through updating a "
            "running Gateway server.",
            error_code=BAD_REQUEST,
        )
    if route_type is _sentinel:
        route_type = None  
    payload = {
        "name": name,
        "route_type": route_type,
        "model": model,
    }
    response = self._call_endpoint(
        "POST", MLFLOW_GATEWAY_CRUD_ROUTE_BASE, json.dumps(payload)
    ).json()
    return Route(**response)
```

Reply:
I don't think the sentinel approach is valid here because `model` is an arg, which appears after `route_type` which is a kwarg. Python doesn't let you define this:

```
SyntaxError: non-default argument follows default argument
```

Reply:
Can we do `route_type=None`, `model=None` and document when these are required?

Reply:
Updated both to be `None` by default, added error checking, updated docs, and added test cases to cover these scenarios.